core/lib/patternlab.js

```diff
@@ -35,18 +36,43 @@ function buildPatternData(dataFilesPath, fsDep) {
 
 // GTP: these two diveSync pattern processors factored out so they can be reused
 // from unit tests to reduce code dupe!
-function processAllPatternsIterative(pattern_assembler, patterns_dir, patternlab) {
-  diveSync(
-    patterns_dir,
-    function (err, file) {
-      //log any errors
-      if (err) {
-        console.log(err);
-        return;
-      }
-      pattern_assembler.process_pattern_iterative(path.relative(patterns_dir, file), patternlab);
-    }
-  );
+function processAllPatternsIterative(pattern_assembler, patterns_dir, patternlab){
+  const async = require('async');
+
+  const promiseAllPatternFiles = new Promise(function (resolve, reject) {
+    dive(
+      patterns_dir,
+      (err, file) => {
+        //log any errors
+        if (err) {
+          console.log('error in processAllPatternsIterative():', err);
+          return;
+        }
+
+        // We now have the loading and process phases spearated; this
+        // loads all the patterns before beginning any analysis, so we
+        // can load them asynchronously and be sure we know about all
+        // of them before we start lineage hunting, for
+        // example. Incidentally, this should also allow people to do
+        // horrifying things like include a page in a atom. But
+        // please, if you're reading this: don't.
+
+        // NOTE: sync for now
+        pattern_assembler.load_pattern_iterative(path.relative(patterns_dir, file), patternlab);
+      },
+      resolve
```

Comment:
Nope, this is the complete handler according to `dive` API: `dive(directory[, options], action[, complete]);`